packages/core/test/acceptance/di_spec.ts

```
@@ -4507,6 +4507,93 @@ describe('di', () => {
+      @Directive({selector: '[dir]', standalone: true})
+      class Dir {
+        value = inject(new HostTagNameToken());
+      }
+
+      @Component({
+        standalone: true,
+        template: `
+          <div dir #v1></div>
+          <span dir #v2></span>
+          <svg dir #v3></svg>
+          <custom-component dir #v4></custom-component>
+          <video dir #v5></video>
+        `,
+        imports: [Dir],
+      })
+      class TestCmp {
+        @ViewChild('v1', {read: Dir}) value1!: Dir;
+        @ViewChild('v2', {read: Dir}) value2!: Dir;
+        @ViewChild('v3', {read: Dir}) value3!: Dir;
+        @ViewChild('v4', {read: Dir}) value4!: Dir;
+        @ViewChild('v5', {read: Dir}) value5!: Dir;
+      }
+
+      const fixture = TestBed.createComponent(TestCmp);
+      fixture.detectChanges();
+      expect(fixture.componentInstance.value1.value).toBe('div');
+      expect(fixture.componentInstance.value2.value).toBe('span');
+      expect(fixture.componentInstance.value3.value).toBe('svg');
+      expect(fixture.componentInstance.value4.value).toBe('custom-component');
+      expect(fixture.componentInstance.value5.value).toBe('video');
+    });
+
+    it('should throw a DI error when injecting into non-DOM nodes', () => {
+      @Directive({selector: '[dir]', standalone: true})
+      class Dir {
+        value = inject(new HostTagNameToken());
+      }
+
+      @Component({
+        standalone: true,
+        template: '<ng-container dir></ng-container>',
+        imports: [Dir],
+      })
+      class TestNgContainer {
+        @ViewChild(Dir) dir!: Dir;
+      }
+
+      @Component({
+        standalone: true,
+        template: '<ng-template dir></ng-template>',
+        imports: [Dir],
+      })
+      class TestNgTemplate {
+        @ViewChild(Dir) dir!: Dir;
+      }
+
+      expect(() => TestBed.createComponent(TestNgContainer))
+          .toThrowError(/No provider for HostTagNameToken found/);
```

I think it'd be great to improve this error message to make it more actionable. The `injectTagName` function would be invoked by the DI code with an argument that represents injection flags. If we are trying to inject a tag name on non-element nodes and there is no "optional" flag specified - we can throw an error and mention which host node is used (and why it doesn't have a tag name) and suggest to use the "optional" flag.